pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go

```diff
@@ -588,3 +645,80 @@ func validateSnapshotStatus(ar *admissionv1.AdmissionRequest, vm *v1.VirtualMach
 
 	return nil
 }
+
+func (admitter *VMsAdmitter) validateVMUpdate(oldVM, newVM *v1.VirtualMachine) []metav1.StatusCause {
+	if newVM.Status.Ready {
+		if !equality.Semantic.DeepEqual(&oldVM.Spec.LiveUpdateFeatures, &newVM.Spec.LiveUpdateFeatures) {
+			return []metav1.StatusCause{{
+				Type:    metav1.CauseTypeFieldValueNotSupported,
+				Message: fmt.Sprintf("Cannot update VM live features while VM is running"),
+				Field:   k8sfield.NewPath("spec").String(),
+			}}
+		}
+
+		if newVM.Spec.LiveUpdateFeatures != nil && newVM.Spec.LiveUpdateFeatures.CPU != nil {
+			oldTopology := oldVM.Spec.Template.Spec.Domain.CPU
+			newTopology := newVM.Spec.Template.Spec.Domain.CPU
+			if oldTopology != nil && newTopology != nil {
+				if oldTopology.Cores != newTopology.Cores {
+					return []metav1.StatusCause{{
+						Type:    metav1.CauseTypeFieldValueNotSupported,
+						Message: fmt.Sprintf("Cannot update CPU cores while live update features configured"),
+						Field:   k8sfield.NewPath("spec.template.spec.domain.cpu.cores").String(),
+					}}
+				}
+				if oldTopology.Threads != newTopology.Threads {
+					return []metav1.StatusCause{{
+						Type:    metav1.CauseTypeFieldValueNotSupported,
+						Message: fmt.Sprintf("Cannot update CPU threads while live update features configured"),
+						Field:   k8sfield.NewPath("spec.template.spec.domain.cpu.cores").String(),
+					}}
+				}
+				if oldTopology.Sockets != newTopology.Sockets {
+					if causeErr, allow := admitter.shouldAllowCPUHotPlug(oldVM); !allow {
+						return []metav1.StatusCause{{
+							Type:    metav1.CauseTypeFieldValueNotSupported,
+							Message: causeErr.Error(),
+							Field:   k8sfield.NewPath("spec.template.spec.domain.cpu.sockets").String(),
+						}}
+					}
+
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (admitter *VMsAdmitter) shouldAllowCPUHotPlug(vm *v1.VirtualMachine) (error, bool) {
+	vmi, err := admitter.VirtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, &metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, true
+	}
```

Comment:
Is it possible that the vmi is not found in this stage? Does it mean we can freely update then?

Reply:
Yes I think we should exist here with error even if the VMI wasn't found because we got here under the condition that VM status is ready, so when its VMI not found I think we should error early as possible.

Reply:
I agree